src/saltext/vmware/modules/license_mgr.py

```diff
@@ -0,0 +1,207 @@
+# Copyright 2021 VMware, Inc.
+# SPDX-License-Identifier: Apache-2.0
+import logging
+
+# noreorder
+import salt.exceptions
+import saltext.vmware.utils.common as utils_common
+import saltext.vmware.utils.license_mgr as utils_license_mgr
+from saltext.vmware.utils.connect import get_service_instance
+
+log = logging.getLogger(__name__)
+
+try:
+    from pyVmomi import vim
+
+    HAS_PYVMOMI = True
+except ImportError:
+    HAS_PYVMOMI = False
+
+
+__virtualname__ = "vmware_license_mgr"
+__proxyenabled__ = ["vmware_license_mgr"]
+__func_alias__ = {"list_": "list"}
+
+
+def __virtual__():
+    if not HAS_PYVMOMI:
+        return False, "Unable to import pyVmomi module."
+    return __virtualname__
+
+
+def add(license_key, **kwargs):
+    """
+    Add a license specified by license key to a Datacenter, Cluster, ESXI Server or vCenter
+    If no datacenter, cluster or ESXI Server is specified, it is assumed the operation is to be applied to a vCenter
+
+    license_key
+        License Key which specifies license to add to license manager
+
+    service_instance
+        Use this vCenter service connection instance instead of creating a new one [default None]
+
+    datacenter_name
+        Datacenter name to use for the operation [default None]
+
+    cluster_name
+        Name of the cluster to add license [default None]
+
+    esxi_hostname
+        Hostname of the ESXI Server to add license [default None]
+
+    CLI Example:
+
+    .. code-block: bash
+
+        salt '*' vmware_license_mgr.add license_key datacenter_name=dc1
+    """
+    log.debug("start vmware ext license_mgr add license_key")
+    ret = {}
+    op = {}
+    for key, value in kwargs.items():
+        op[key] = value
+
+    service_instance = op.pop("service_instance", None)
+    datacenter_name = op.pop("datacenter_name", None)
+    cluster_name = op.pop("cluster_name", None)
+    esxi_hostname = op.pop("esxi_hostname", None)
+
+    if service_instance is None:
+        service_instance = get_service_instance(opts=__opts__, pillar=__pillar__)
+
+    if not utils_license_mgr.is_vcenter(service_instance):
+        ret["message"] = "Failed, not connected to a vCenter"
+        ret["result"] = False
+        return ret
+
+    try:
+        if "test" in __opts__ and __opts__["test"]:
+            ret["licenses"] = license_key
```

Comment:
since the same response is used for state, state module returns, `comment`, `changes` and `result` only. Also, haven't seen module act as a state too before, not sure if that matters. @waynew may have some insights.

Reply:
For example from salt/modules/junos.py
2299     ret_messages = ""
2300     if not source.startswith("/"):
2301         ret["message"] = "Source directory must be a fully qualified path."
2302         ret["success"] = False
2303         return ret

so maintaining the same coding style

Reply:
👍 Resolving conversation based on quoted salt code pattern.